### DISCUSSION TITLE
Did I see that Wifi on the rp2040 is coming?

### DISCUSSION THREAD

**Discussion by marklacas** [tabemann/zeptoforth#22] [Q&A]
Hoping to use wifi at some point.
Thanks,
-ml

**Comment by tabemann** [comment]
While there is a Pico-W, I will probably have WiFi working first for the WIO RP2040 because its ESP radio functionality abstracts much of the protocol stack away, whereas it seems to me that on the Pico-W much more work is needed to get WiFi and IP working due to requiring an actual protocol stack on the RP2040 itself.

**Reply by znmeb** [reply]
The C and MicroPython code for wireless on the Pico W is still a work in progress. There are some working examples but it's a bit bleeding edge still.

**Comment by rodgetk** [comment]
I'm also interested in the Pico wifi :-)
Thanks,
/Roger

**Reply by tabemann** [reply]
It is in the works, but WiFi support for the Wio RP2040 will come before support for the Pico W.

**Reply by znmeb** [reply]
Is the Pico W wireless on the roadmap now?

**Reply by tabemann** [reply]
Yes it is! There is a CYW43439 driver (for WiFi) and an IP stack in `extra/rp2040` in the `devel` branch now. I do currently consider it to be beta software, hence why I have not put it in the `master` branch, but I would be very glad if you tried it out. Make sure to read [this guide](https://github.com/tabemann/zeptoforth/blob/devel/BUILDING_AND_USING_ZEPTOIP.md) prior to using it. Note that this requires non-free firmware for the CYW43439 which is kept [here](https://github.com/tabemann/cyw43-firmware) and which you will need to upload to the Pico W yourself. Also, check out the [test programs](https://github.com/tabemann/zeptoforth/tree/devel/test/rp2040) marked `pico_w_net_*`.

**Reply by znmeb** [reply]
Thanks!! I'm getting a new board next week (https://shop.pimoroni.com/products/picovision?variant=41048911904851) and I'll be doing some extensive testing.

**Reply by tabemann** [reply]
Note that you need not recompile zeptoforth as a whole to use the IP stack (which I call zeptoIP) or the CYW43439 driver - the rp2040_big binaries in the latest release tarball will work with it.

**Comment by znmeb** [comment]
Sounds good. The device came early - it's actually here now and an HDMI cable is on its way. I need to verify that the MicroPython and C/C++ toolchains work first, but then I can put Zeptoforth on it. So it'll probably be Sunday before I test Zeptoforth.

**Reply by tabemann** [reply]
I should note that you will have to implement a good amount of functionality yourself to interface with this board which is likely already implemented in the version of MicroPython for this board. You will need to implement the interface with the "GPU" on top of I2C, and it seems that you will need SPI to talk to the PSRAM (I saw something online about Infineon's PSRAM using something called "xSPI", which implies that you may need to use the PIO to implement the protocol).

**Comment by tabemann** [comment]
One thing I should note is, despite my earlier comment in this thread, Wio RP2040 WiFi support is permanently shelved, for the reason that I just could not get it reliable enough for me to justify formally releasing the code.

**Reply by znmeb** [reply]
What were the issues? Connection drops? That could well be a hardware problem - the Pico W may not have the same radio stability as the ESP32 ones that have been around a lot longer. I do have one of those Arduino boards with both an RP2040 and an ESP32 where the ESP32 does the wifi. But I don't have any radio signal strength measurement tools besides my phone.

**Reply by tabemann** [reply]
The issues were with the communication protocol, with was a packetized text-based AT protocol, between the RP2040 and the ESP8285 ─ to get it to work at all I had to add large numbers of little delays and critical sections throughout the code, which hurt its performance greatly, yet even then it would frequently get confused when the RP2040 and the ESP8285 got out of sync, and transmission would lock up then. I did stress tests where I constantly sent data from the Wio RP2040 to my laptop over WiFi, and not only was it slow, but it would only run for generally only a few hours tops before locking up (well, I did see it run for about eight hours on a few occasions, but they still locked up).

**Reply by tabemann** [reply]
I should also note that I had trouble getting to work with the official MicroPython binary, as did other people who tried using the board (it got generally bad reviews from what I saw). They did their best to hide the source code for the MicroPython version for their board, and when I found it and tried to implement my own code based off of it, once I managed to get my code sort-of working I wondered if they had even tested their code, as it seemed that here was no way it could have worked.

**Reply by znmeb** [reply]
I just got a Pimoroni PicoVision today and they have their own MicroPython build on GitHub. It has a Pico W on board and if I have WiFi / Bluetooth issues with it I can open issues there. Here's the repo if you want to follow along: https://github.com/pimoroni/picovision/releases

**Comment by tabemann** [comment]
It has been a good while since I got zeptoIP for IPv4 working, so I am closing this thread.

Edit: accidentally wrote "zeptoforth" rather than "zeptoIP".
